sktime/forecasting/compose/_fallback.py

```diff
@@ -0,0 +1,171 @@
+a robust forecasting mechanism by providing fallback options.
+"""
+
+__author__ = ["ninedigits"]
+__all__ = ["FallbackForecaster"]
+
+import warnings
+
+from sktime.base import _HeterogenousMetaEstimator
+from sktime.forecasting.base import BaseForecaster
+
+
+class FallbackForecaster(_HeterogenousMetaEstimator, BaseForecaster):
+    """Forecaster that sequentially tries a list of forecasting models.
+
+    Attempts to fit the provided forecasters in the order they are given. If a
+    forecaster fails during fitting or prediction, it proceeds to the next one. This
+    class is useful in scenarios where the reliability of individual forecasting models
+    may be in question, and a fallback mechanism is desired.
+
+    Parameters
+    ----------
+    forecasters : list of (str, estimator) tuples
+        Forecasters to be tried sequentially.
+    warn : bool, default=False
+        If True, raises warnings when a forecaster fails to fit or predict.
+
+    Examples
+    --------
+    >>> from sktime.forecasting.naive import NaiveForecaster
+    >>> from sktime.forecasting.compose import FallbackForecaster
+    >>> from sktime.forecasting.compose import EnsembleForecaster
+    >>> from sktime.forecasting.trend import PolynomialTrendForecaster
+    >>> from sktime.datasets import load_airline
+    >>> y = load_airline()
+    >>> forecasters = [
+    ...     (
+    ...         "ensemble",
+    ...         EnsembleForecaster(
+    ...             [
+    ...                 ("trend", PolynomialTrendForecaster()),
+    ...                 ("polynomial", NaiveForecaster())
+    ...             ]
+    ...         )
+    ...     ),
+    ...     ("naive", NaiveForecaster())
+    ... ]
+    >>> forecaster = FallbackForecaster(forecasters=forecasters)
+    >>> forecaster.fit(y=y, fh=[1, 2, 3])
+    FallbackForecaster(...)
+    >>> y_pred = forecaster.predict()
+    """
+
+    def __init__(self, forecasters, warn=False):
+        super().__init__()
+        self.forecasters = tuple(forecasters)
+        self.current_forecaster = None
+        self.current_name = None
+        self.warn = warn
+        self._set_forecast_pointer(0)
+
+        self._anytagis_then_set("requires-fh-in-fit", True, False, self.forecasters)
+
+    def _set_forecast_pointer(self, val):
+        """Set the current forecaster index to a specified value.
+
+        This method ensures the persistence of `__current_forecaster_index` across the
+        invocations of `.fit()`. Typically, calling `.fit()` invokes `.reset()`, which,
+        by default, would erase certain instance variables, including
+        `__current_forecaster_index`. By using this method, `__current_forecaster_index`
+        is preserved or initialized when necessary.
+
+        Parameters
+        ----------
+        val : int
+            The new value to which `__current_forecaster_index` should be set.
+
+        Returns
+        -------
+        None
+        """
+        try:
```

Comment:
@fkiraly What do you think about this method of keeping track of the current forecasting index?

Reply:
* I would use a variable ending in underscore, such as `first_nonfailing_forecaster_index_`, this can then be read using `get_fitted_params`
* I would also save the exceptions to a dict in `_fit`, e.g., `exceptions_raised_`, instead of raising warnings. Since this forecaster expects failures, it is perhaps better to return the exact failures so users could handle them more easily.

Reply:
@fkiraly 
> I would write an attr ending in underscore in _fit to self, such as first_nonfailing_forecaster_index_, this can then be read using get_fitted_params

So I tried just about every variation of this, but the only way I could get it to work correctly is to use a double underscore. The issue is that if `fit` succeeds but then fails at `predict`, `.fit` is called again within predict, which triggers the `reset` method. 

https://github.com/sktime/sktime/blob/a833f8968f3a87e0d5a12571e40ded22bca0109b/sktime/forecasting/base/_base.py#L348-L350

Within `reset`, my pointers are erased unless they have a double underscore.

```python3
    def reset(self):
        ...
        params = self.get_params(deep=False)

        # delete all object attributes in self
        attrs = [attr for attr in dir(self) if "__" not in attr]
        cls_attrs = list(dir(type(self)))
        self_attrs = set(attrs).difference(cls_attrs)
        for attr in self_attrs:
            delattr(self, attr)
        ...
```
But I can't call `self.get_fitted_params` unless I've already fit the estimator. The workaround I did was to use the double underscore to avoid the deletion in `reset`, and then to use `getattr` to get the value or set the value to the initial state.

```python3
    def _fit(self, y, X=None, fh=None):
        ...
        self.first_nonfailing_forecaster_index__ = getattr(
            self, "first_nonfailing_forecaster_index__", 0
        )
        self.exceptions_raised__ = getattr(self, "exceptions_raised__", dict())
        ...
```

Reply:
I don't get this, is it no the inner `fit` you are calling in `predict`?

Can you be more detailed on which object's attributes are erased when?

Reply:
@fkiraly 
The issue is caused when `self.fit()` is called within `self._predict()`, which occurs if a fit is successful but the forecaster fails at predict.  Here's an example.
```python3
 def test_fallbackforecaster_fails_twice_simple():
    """First two FallbackForecasters fail, third succeeds"""
    y = make_forecasting_problem(random_state=42)
    forecaster1 = DummyForecaster(raise_at="fit")
    forecaster2 = DummyForecaster(raise_at="predict")
    forecaster3 = PolynomialTrendForecaster()
    forecaster4 = NaiveForecaster()

    forecaster = FallbackForecaster(
        [
            ("forecaster1_fails_fit", forecaster1),
            ("forecaster2_fails_prd", forecaster2),
            ("forecaster3_succeeded", forecaster3), # Expected forecaster
            ("forecaster4_notcalled", forecaster4),
        ]
    )
    forecaster.fit(y=y, fh=[1, 2, 3])
    y_pred_actual = forecaster.predict() # Will fail at first predict 
    # and have to call fit again, erasing pointer and dict. Screenshot attached shows deletion step


```
<img width="1040" alt="Screenshot 2024-01-23 at 10 57 18 AM" src="https://github.com/sktime/sktime/assets/16393653/e31d61a4-406e-4e2f-b915-2f033ed6afe1">

It's entirely possible that I'm missing something obvious that I needed to include in the code to avoid this, but this is what immediately stands out to me.

Reply:
oh, I see! The solution afaik is not to fit `self`, which leads to recursion, but only a clone of the forecaster that the `first_nonfailing_forecaster_index_` points to. Would that work?

Reply:
That works, although I don't necessarily need to clone the forecaster, I can just call forecaster.fit(). Side note, is there a way for me to run these checks locally? 
<img width="300" alt="image" src="https://github.com/sktime/sktime/assets/16393653/3ec3dca3-e522-4c61-854e-542390c24df5">

Reply:
> Side note, is there a way for me to run these checks locally?

yes, use `sktime.utils.check_estimator`, that will run them only for your forecaster.

Or, to run all of them, you need to set up `pytest`, though that is only needed typically if you are contributing to the framework:
https://www.sktime.net/en/stable/developer_guide/continuous_integration.html

> although I don't necessarily need to clone the forecaster, I can just call forecaster.fit

I think you do, otherwise you will fail the "does't change `__init__` params" test?

Reply:
Got it, that explains the errors. I failed a bunch of those in the previous commit.

Reply:
I was able to run all the pytests and have them pass locally on my computer. Crossing my fingers.